From 90c677e03718a123c099d504eab7c4edfba90779 Mon Sep 17 00:00:00 2001
From: Quentin Casasnovas <quentin.casasnovas@vates.tech>
Date: Tue, 17 Mar 2026 09:21:49 +0100
Subject: [PATCH] !kabi writeback: synchronize sync(2) against cgroup writeback
 membership switches

We already were using a shadow variable for this struct for the dev_name
field - instead of a creating a shadow of a different type, just add the
two new fields to a shared struct as this simplifies greatly the lifetime
management of the shadow struct and error paths.

Fixes: edca54b897bb ("writeback: synchronize sync(2) against cgroup writeback membership switches")
Signed-off-by: Quentin Casasnovas <quentin.casasnovas@vates.tech>
---
 fs/fs-writeback.c                | 33 ++++++++++++++++--------
 include/linux/backing-dev-defs.h | 13 +++++++++-
 mm/backing-dev.c                 | 43 +++++++++++++++++++++-----------
 3 files changed, 64 insertions(+), 25 deletions(-)

diff --git a/fs/fs-writeback.c b/fs/fs-writeback.c
index 4b6d7908acf4..0f726cd02ff8 100644
--- a/fs/fs-writeback.c
+++ b/fs/fs-writeback.c
@@ -13,6 +13,7 @@
  *		Additions for address_space-based writeback
  */
 
+#include "linux/backing-dev-defs.h"
 #include <linux/kernel.h>
 #include <linux/export.h>
 #include <linux/spinlock.h>
@@ -30,6 +31,8 @@
 #include <linux/memcontrol.h>
 #include "internal.h"
 
+#include <linux/shadow_var.h>
+
 /*
  * 4MB minimal write chunk size
  */
@@ -333,14 +336,18 @@ struct inode_switch_wbs_context {
 	struct work_struct	work;
 };
 
-static void bdi_down_write_wb_switch_rwsem(struct backing_dev_info *bdi)
+static void bdi_down_write_wb_switch_rwsem(struct kabi_shadow_backing_dev_info *shadow_bdi)
 {
-	down_write(&bdi->wb_switch_rwsem);
+	if (unlikely(shadow_bdi == NULL))
+		return;
+	down_write(&shadow_bdi->wb_switch_rwsem);
 }
 
-static void bdi_up_write_wb_switch_rwsem(struct backing_dev_info *bdi)
+static void bdi_up_write_wb_switch_rwsem(struct kabi_shadow_backing_dev_info *shadow_bdi)
 {
-	up_write(&bdi->wb_switch_rwsem);
+	if (unlikely(shadow_bdi == NULL))
+		return;
+	up_write(&shadow_bdi->wb_switch_rwsem);
 }
 
 static void inode_switch_wbs_work_fn(struct work_struct *work)
@@ -349,6 +356,7 @@ static void inode_switch_wbs_work_fn(struct work_struct *work)
 		container_of(work, struct inode_switch_wbs_context, work);
 	struct inode *inode = isw->inode;
 	struct backing_dev_info *bdi = inode_to_bdi(inode);
+	struct kabi_shadow_backing_dev_info *shadow_bdi = shadow_var_get(bdi, "kabi_shadow_backing_dev_info");
 	struct address_space *mapping = inode->i_mapping;
 	struct bdi_writeback *old_wb = inode->i_wb;
 	struct bdi_writeback *new_wb = isw->new_wb;
@@ -360,7 +368,8 @@ static void inode_switch_wbs_work_fn(struct work_struct *work)
 	 * If @inode switches cgwb membership while sync_inodes_sb() is
 	 * being issued, sync_inodes_sb() might miss it.  Synchronize.
 	 */
-	down_read(&bdi->wb_switch_rwsem);
+	if (shadow_bdi)
+		down_read(&shadow_bdi->wb_switch_rwsem);
 
 	/*
 	 * By the time control reaches here, RCU grace period has passed
@@ -454,7 +463,8 @@ static void inode_switch_wbs_work_fn(struct work_struct *work)
 	spin_unlock(&new_wb->list_lock);
 	spin_unlock(&old_wb->list_lock);
 
-	up_read(&bdi->wb_switch_rwsem);
+	if (shadow_bdi)
+		up_read(&shadow_bdi->wb_switch_rwsem);
 
 	if (switched) {
 		wb_wakeup(new_wb);
@@ -489,6 +499,7 @@ static void inode_switch_wbs_rcu_fn(struct rcu_head *rcu_head)
 static void inode_switch_wbs(struct inode *inode, int new_wb_id)
 {
 	struct backing_dev_info *bdi = inode_to_bdi(inode);
+	struct kabi_shadow_backing_dev_info *shadow_bdi = shadow_var_get(bdi, "kabi_shadow_backing_dev_info");
 	struct cgroup_subsys_state *memcg_css;
 	struct inode_switch_wbs_context *isw;
 
@@ -502,7 +513,7 @@ static void inode_switch_wbs(struct inode *inode, int new_wb_id)
 	 * blocks heavily, we might end up starting a large number of
 	 * switches which will block on the rwsem.
 	 */
-	if (!down_read_trylock(&bdi->wb_switch_rwsem))
+	if (shadow_bdi && !down_read_trylock(&shadow_bdi->wb_switch_rwsem))
 		return;
 
 	isw = kzalloc(sizeof(*isw), GFP_ATOMIC);
@@ -554,7 +565,8 @@ static void inode_switch_wbs(struct inode *inode, int new_wb_id)
 		wb_put(isw->new_wb);
 	kfree(isw);
 out_unlock:
-	up_read(&bdi->wb_switch_rwsem);
+	if (shadow_bdi)
+		up_read(&shadow_bdi->wb_switch_rwsem);
 }
 
 /**
@@ -2453,6 +2465,7 @@ void sync_inodes_sb(struct super_block *sb)
 		.for_sync	= 1,
 	};
 	struct backing_dev_info *bdi = sb->s_bdi;
+	struct kabi_shadow_backing_dev_info *shadow_bdi = shadow_var_get(bdi, "kabi_shadow_backing_dev_info");
 
 	/*
 	 * Can't skip on !bdi_has_dirty() because we should wait for !dirty
@@ -2464,10 +2477,10 @@ void sync_inodes_sb(struct super_block *sb)
 	WARN_ON(!rwsem_is_locked(&sb->s_umount));
 
 	/* protect against inode wb switch, see inode_switch_wbs_work_fn() */
-	bdi_down_write_wb_switch_rwsem(bdi);
+	bdi_down_write_wb_switch_rwsem(shadow_bdi);
 	bdi_split_work_to_wbs(bdi, &work, false);
 	wb_wait_for_completion(bdi, &done);
-	bdi_up_write_wb_switch_rwsem(bdi);
+	bdi_up_write_wb_switch_rwsem(shadow_bdi);
 
 	wait_sb_inodes(sb);
 }
diff --git a/include/linux/backing-dev-defs.h b/include/linux/backing-dev-defs.h
index e9d66f77062b..e68abbc2db80 100644
--- a/include/linux/backing-dev-defs.h
+++ b/include/linux/backing-dev-defs.h
@@ -190,7 +190,13 @@ struct backing_dev_info {
 	struct radix_tree_root cgwb_tree; /* radix tree of active cgroup wbs */
 	struct rb_root cgwb_congested_tree; /* their congested states */
 	struct mutex cgwb_release_mutex;  /* protect shutdown of wb structs */
-	struct rw_semaphore wb_switch_rwsem; /* no cgwb switch while syncing */
+	/**
+	 * Added in edca54b897bb ("writeback: synchronize sync(2) against
+	 * cgroup writeback membership switches"), converted to use shadow
+	 * structs.
+	 *
+	 * struct rw_semaphore wb_switch_rwsem; * no cgwb switch while syncing *
+	 */
 #else
 	struct bdi_writeback_congested *wb_congested;
 #endif
@@ -214,6 +220,11 @@ struct backing_dev_info {
 #endif
 };
 
+struct kabi_shadow_backing_dev_info {
+	char dev_name[64];
+	struct rw_semaphore wb_switch_rwsem; /* no cgwb switch while syncing */
+};
+
 enum {
 	BLK_RW_ASYNC	= 0,
 	BLK_RW_SYNC	= 1,
diff --git a/mm/backing-dev.c b/mm/backing-dev.c
index 69b303bbc639..2919b72b9235 100644
--- a/mm/backing-dev.c
+++ b/mm/backing-dev.c
@@ -1,4 +1,5 @@
 
+#include "linux/gfp.h"
 #include <linux/wait.h>
 #include <linux/backing-dev.h>
 #include <linux/kthread.h>
@@ -252,10 +253,18 @@ static int __init default_bdi_init(void)
 {
 	int err;
 
+	if (!shadow_var_alloc(&noop_backing_dev_info, "kabi_shadow_backing_dev_info", sizeof(struct kabi_shadow_backing_dev_info), GFP_KERNEL)) {
+		return -ENOMEM;
+	}
+
+
 	bdi_wq = alloc_workqueue("writeback", WQ_MEM_RECLAIM | WQ_UNBOUND |
 				 WQ_SYSFS, 0);
-	if (!bdi_wq)
+	if (!bdi_wq) {
+		shadow_var_free(&noop_backing_dev_info, "kabi_shadow_backing_dev_info");
 		return -ENOMEM;
+	}
+
 
 	err = bdi_init(&noop_backing_dev_info);
 
@@ -687,12 +696,18 @@ struct bdi_writeback *wb_get_create(struct backing_dev_info *bdi,
 
 static int cgwb_bdi_init(struct backing_dev_info *bdi)
 {
+	struct kabi_shadow_backing_dev_info *shadow_bdi;
 	int ret;
 
+	shadow_bdi = shadow_var_get(bdi, "kabi_shadow_backing_dev_info");
+	if (!shadow_bdi)
+		return -ENOMEM;
+
+
 	INIT_RADIX_TREE(&bdi->cgwb_tree, GFP_ATOMIC);
 	bdi->cgwb_congested_tree = RB_ROOT;
 	mutex_init(&bdi->cgwb_release_mutex);
-	init_rwsem(&bdi->wb_switch_rwsem);
+	init_rwsem(&shadow_bdi->wb_switch_rwsem);
 
 	ret = wb_init(&bdi->wb, bdi, 1, GFP_KERNEL);
 	if (!ret) {
@@ -868,13 +883,13 @@ struct backing_dev_info *bdi_alloc_node(gfp_t gfp_mask, int node_id)
 		return NULL;
 
 
-	if (!shadow_var_alloc(bdi, "dev_name", sizeof(char[64]), gfp_mask)) {
+	if (!shadow_var_alloc(bdi, "kabi_shadow_backing_dev_info", sizeof(struct kabi_shadow_backing_dev_info), gfp_mask)) {
 		kfree(bdi);
 		return NULL;
 	}
 
 	if (bdi_init(bdi)) {
-		shadow_var_free(bdi, "dev_name");
+		shadow_var_free(bdi, "kabi_shadow_backing_dev_info");
 		kfree(bdi);
 		return NULL;
 	}
@@ -885,17 +900,17 @@ EXPORT_SYMBOL(bdi_alloc_node);
 int bdi_register_va(struct backing_dev_info *bdi, const char *fmt, va_list args)
 {
 	struct device *dev;
-	char (*kabi_bdi_dev_name)[64];
+	struct kabi_shadow_backing_dev_info *shadow_bdi;
 
 	if (bdi->dev)	/* The driver needs to use separate queues per device */
 		return 0;
 
-	kabi_bdi_dev_name = shadow_var_get(bdi, "dev_name");
-	if (!kabi_bdi_dev_name)
+	shadow_bdi = shadow_var_get(bdi, "kabi_shadow_backing_dev_info");
+	if (!shadow_bdi)
 		return -ENOMEM;
 
-	vsnprintf(*kabi_bdi_dev_name, sizeof(*kabi_bdi_dev_name), fmt, args);
-	dev = device_create(bdi_class, NULL, MKDEV(0, 0), bdi, *kabi_bdi_dev_name);
+	vsnprintf(shadow_bdi->dev_name, sizeof(shadow_bdi->dev_name), fmt, args);
+	dev = device_create(bdi_class, NULL, MKDEV(0, 0), bdi, shadow_bdi->dev_name);
 	if (IS_ERR(dev))
 		return PTR_ERR(dev);
 
@@ -989,7 +1004,7 @@ static void release_bdi(struct kref *ref)
 	WARN_ON_ONCE(bdi->dev);
 	wb_exit(&bdi->wb);
 	cgwb_bdi_exit(bdi);
-	shadow_var_free(bdi, "dev_name");
+	shadow_var_free(bdi, "kabi_shadow_backing_dev_info");
 	kfree(bdi);
 }
 
@@ -1001,16 +1016,16 @@ EXPORT_SYMBOL(bdi_put);
 
 const char *bdi_dev_name(struct backing_dev_info *bdi)
 {
-	char (*kabi_bdi_dev_name)[64];
+	struct kabi_shadow_backing_dev_info *shadow_bdi;
 
 	if (!bdi || !bdi->dev)
 		return bdi_unknown_name;
 
-	kabi_bdi_dev_name = shadow_var_get(bdi, "dev_name");
-	if (kabi_bdi_dev_name == NULL)
+	shadow_bdi = shadow_var_get(bdi, "dev_name");
+	if (shadow_bdi == NULL)
 		return bdi_unknown_name;
 
-	return *kabi_bdi_dev_name;
+	return shadow_bdi->dev_name;
 }
 EXPORT_SYMBOL_GPL(bdi_dev_name);
 
-- 
2.43.0

